### PR TITLE
Use dotnet-coverage instead of CodeCoverage.exe

### LINKED ISSUE #4745: Replace CodeCoverage.exe by dotnet-coverage global tool or Microsoft.CodeCoverage.Console
# Feature Request

replace the way how code coverage is executed.

**Is your feature request related to a problem?**

From CodeCoverage execution with Microsoft.CodeCoverage 17.7.0

```log
CodeCoverage.exe is deprecated. Use dotnet-coverage global tool or Microsoft.CodeCoverage.Console instead. Some features of CodeCoverage.exe will be removed in next release of Visual Studio.
```

## Additional Context

While working on this, propagate solution to https://github.com/open-telemetry/opentelemetry-dotnet-contrib repository

### PR DESCRIPTION
**Pull request by Kielek**
Fixes #4745

## Changes

* Use dotnet-coverage instead of CodeCoverage.exe
* Updated test dedepnecnes
* Be less strict about test dependencies

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests ~~added/updated~~ result processing updated
* ~~[ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes~~
* ~~[ ] Changes in public API reviewed (if applicable)~~

## Results

Results from changes steps contains more data now, but the main part is similar. Results for random method:

befoer
```xml
        <function id="12547" token="0x6000052" name="&lt;AddOpenTelemetry_WithLogging_HostConfigurationHonoredTest&gt;b__12_0(Microsoft.Extensions.Configuration.IConfigurationBuilder)" namespace="OpenTelemetry.Extensions.Hosting.Tests" type_name="OpenTelemetryServicesExtensionsTests.&lt;&gt;c" block_coverage="100.00" line_coverage="100.00" blocks_covered="4" blocks_not_covered="0" lines_covered="5" lines_partially_covered="0" lines_not_covered="0">
          <ranges>
            <range source_id="5" covered="yes" start_line="390" start_column="17" end_line="393" end_column="20" />
            <range source_id="5" covered="yes" start_line="394" start_column="13" end_line="394" end_column="14" />
          </ranges>
        </function>
		
```

after

```xml
        <function block_coverage="100.00" line_coverage="100.00" blocks_covered="4" blocks_not_covered="0" lines_covered="5" lines_partially_covered="0" lines_not_covered="0" id="14686" token="0x6000069" name="&lt;AddOpenTelemetry_WithLogging_HostConfigurationHonoredTest&gt;b__12_0(Microsoft.Extensions.Configuration.IConfigurationBuilder)" namespace="OpenTelemetry.Extensions.Hosting.Tests" type_name="OpenTelemetryServicesExtensionsTests.&lt;&gt;c">
          <ranges>
            <range source_id="5" covered="yes" start_line="390" start_column="17" end_line="393" end_column="20" />
            <range source_id="5" covered="yes" start_line="394" start_column="13" end_line="394" end_column="14" />
          </ranges>
        </function>
```

### CODE REVIEW

**Comment by Kielek** [comment]
Failed tests seems to be not related to changes. Please rerun CI.

File: .github/workflows/code-coverage.yml [hunk 1]
```diff
@@ -39,6 +39,9 @@ jobs:
     - name: dotnet test
       run: dotnet test --collect:"Code Coverage" --results-directory:"TestResults" --configuration Release --no-build -- RunConfiguration.DisableAppDomain=true
 
+    - name: Install coverage tool
+      run: dotnet tool install -g dotnet-coverage
```
Comment (CodeBlanch) [inline .github/workflows/code-coverage.yml:43]: Funny I have been messing with this too over on https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1347. I think we can go deeper with the changes:

* Remove `.\build\process-codecoverage.ps1`
* Remove `reportgenerator` completely

It seems `dotnet-coverage merge` is able to do all that work. Here is what I have so far on contrib:

```yaml
    - name: Setup dotnet
      uses: actions/setup-dotnet@v3

    - name: Install dependencies
      run: dotnet restore

    - name: dotnet build
      run: dotnet build --configuration Release --no-restore

    - name: dotnet test
      run: dotnet test --collect:"Code Coverage" --results-directory:"TestResults" --configuration Release --no-restore --no-build -- RunConfiguration.DisableAppDomain=true

    - name: Install coverage tool
      run: dotnet tool install -g dotnet-coverage

    - name: Merging test results
      run: dotnet-coverage merge -r -f cobertura -o ./TestResults/Cobertura.xml ./TestResults/*.coverage

    - uses: codecov/codecov-action@v3.1.4
      env:
        OS: ${{ matrix.os }}
      with:
        file: TestResults/Cobertura.xml
        env_vars: OS
        flags: unittests
        name: Code Coverage for ${{ matrix.os }}
```

PS: Note the addition of `env:` in the upload. I don't think the current `env_vars` has ever worked without that.

File: Directory.Packages.props [hunk 2]
```diff
@@ -68,14 +68,14 @@
     <PackageVersion Include="Microsoft.AspNetCore.TestHost" Version="7.0.9" />
     <PackageVersion Include="Microsoft.AspNetCore.Mvc.Testing" Version="7.0.9" />
     <PackageVersion Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="[3.11.0-beta1.23402.2]" />
-    <PackageVersion Include="Microsoft.CodeCoverage" Version="[17.6.3]" />
+    <PackageVersion Include="Microsoft.CodeCoverage" Version="[17.7.2,18.0.0)" />
```
Comment (CodeBlanch) [inline Directory.Packages.props:71]: What I noticed working on https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1347 is that `Microsoft.NET.Test.Sdk` now brings in `Microsoft.CodeCoverage`. What I think we should do is...

* Remove this line: https://github.com/open-telemetry/opentelemetry-dotnet/blob/fe78453c03feb8dbe506b2a0284312bdfa1367c5/build/Common.props#L32
  Kind of strange we were ever adding this to all projects in the first place 🤣 
* Remove the version specified in this props file (it shouldn't be needed).

I have those two changes on my contrib PR and it seems to be working fine.

Comment (Kielek) [reply Directory.Packages.props]: I have executed locally all step, looks good after changes: 2c20e9a110e70ce745aff6561f7e2fac56f6812e

File: .github/workflows/code-coverage.yml (hunk 1 above)
Comment (Kielek) [reply .github/workflows/code-coverage.yml:40]: To be honest I didn't checked your PR in contrib. It was in draft state :)

Suggestions Aoplied.

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Review by utpilla — APPROVED** [review]

**Pull request merged**